### ISSUE TITLE
Help Wanted! How to plot a A 2D color map with color scale ? Similar to QCustomPlot 

### ISSUE THREAD

**Issue by kevinNejad** [pyqtgraph/pyqtgraph#1261]
Hi, I have been trying for couple of days to plot something like a 2D Color Map like [https://www.qcustomplot.com/index.php/demos/colormapdemo](url) I did try to install any library suggested, also tried some libraries which don't seem to be maintained anymore.
I really appreciate if someone guide me on this.
More specifically, I'd like to plot something that I'd do following if I was using matplotlib. 
The purpose of the visualisation is to show the training of a Neural Net. So it needs to be updated after each step, that's why I need a native library.
```
plt.subplot(1,3,3)
xx = np.linspace(-2,2,100)
X,Y = np.meshgrid(xx,xx)
Z = np.vstack([X.reshape(-1),Y.reshape(-1)])
out = NN(torch.tensor(Z.T).float())

out=out.reshape(100,100)
plt.pcolor(xx,xx,out.data)

plt.title('Output')
plt.show()
```

### Tested environment(s)

 * PyQtGraph version:0.11.0
 * Qt Python binding: PyQt5 5.10.1 Qt 5.9.5
 * Python version: 3.6.5
 * NumPy version:1.18.5
 * Operating system: Linux Ubuntu 18.04 LTS
 * Installation method: pip

**Comment by 2xB** [comment]
Are you helped with the examples `pyqtgraph/examples/ImageView.py` and `pyqtgraph/examples/DataSlicing.py`? The former demonstrates an `ImageView` while the latter example demonstrates two `ImageView`s where the lower one is updated based on user interaction on the top one.

**Comment by kevinNejad** [comment]
I ended up using `ImageItem`. However, the issue that I encounter is now updating it's values. I created signal which carries the data(image as 2d array) to update. But it can't emit the signal in a for loop. I'm following MVC design as follows:

```import numpy as np
from PyQt5.QtCore import pyqtSignal, QObject

class Model(QObject):
    def __init__(self):
        QObject.__init__(self)  

    signal= pyqtSignal(np.ndarray)
    def train(self):
        for step in range(1000):
            dummy_data = np.random.normal(size=(100,100))
            self.signal.emit(dummy_data)
```

```
from model import Model
from view import View
import numpy as np
import sys

from PyQt5 import QtCore, QtGui, QtWidgets



class Controller:
    def __init__(self):
        self.model = Model()
        self.view = View(self)

if __name__=='__main__':

    ctrl = Controller()

    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    
    ui = ctrl.view
    ui.setupUi(MainWindow)
    MainWindow.show()

    # w = ClassificationPlot(ui.TaskPlot)
    ctrl.model.signal.connect(ui.update)
    
    ctrl.model.train()
    # It does work when I uncomment below, but it besically call train (with a for loop) everytime and then update
    # t = QtCore.QTimer()
    # t.timeout.connect(ctrl.model.train)
    # t.start(1)

    sys.exit(app.exec_())
```
from PyQt5 import QtCore, QtGui, QtWidgets
from pyqtgraph import PlotWidget
import pyqtgraph as pg
from matplotlib import cm
import numpy as np

class View(object):

    def __init__(self, controller):
        self.controller = controller

    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(640, 480)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.gridLayout = QtWidgets.QGridLayout(self.centralwidget)
        self.gridLayout.setObjectName("gridLayout")

        self.graphicsView = PlotWidget(self.centralwidget)
        self.graphicsView.setObjectName("graphicsView")
        self.gridLayout.addWidget(self.graphicsView, 0, 0, 1, 1)

        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 640, 31))
        self.menubar.setObjectName("menubar")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)
        self.toolBar = QtWidgets.QToolBar(MainWindow)
        self.toolBar.setObjectName("toolBar")
        MainWindow.addToolBar(QtCore.Qt.TopToolBarArea, self.toolBar)

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

        self.img = pg.ImageItem()
        self.graphicsView.addItem(self.img)

        colormap = cm.get_cmap("jet")  # cm.get_cmap("CMRmap")
        colormap._init()
        lut = (colormap._lut * 255).view(np.ndarray)  # Convert matplotlib colormap from 0-1 to 0 -255 for Qt


        self.img.setLookupTable(lut)
        self.img.setLevels([-50,40])


        self.graphicsView.setLabel('left', 'X1')
        self.graphicsView.setLabel('bottom', 'X2')
        
    def update(self, data):
        self.img.setImage(data, levels=(0, 1))
        self.graphicsView.show()

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.toolBar.setWindowTitle(_translate("MainWindow", "toolBar"))

```

Basically, the train function start training a neural network, and while it's training, I need to show how network is being trained. 
I'm pretty new with python GUI, I really appreciate your help.

**Comment by j9ac9k** [comment]
you may be running into some naming collisions issues; on some of my projects I've ran into something similar.  I would recommend changing

```python
class Model(QObject):
    def __init__(self):
        QObject.__init__(self)  

    signal= pyqtSignal(np.ndarray)
    def train(self):
        for step in range(1000):
            dummy_data = np.random.normal(size=(100,100))
            self.signal.emit(dummy_data)
```

to 

```python
class Model(QObject):

    sigNewData = pyqtSignal(np.ndarray)
    
    def __init__(self):
        super().__init__()

    def train(self):
        for step in range(1_000):
            dummy_data = np.random.normal(size=(100, 00))
            self.sigNewData.emit(dummy_data)
```

If you're not sure if the signal is firing, you can also attach the signal to a print statement by adding the line

```python
self.sigNewData.connect(lambda x: print("signNewData signal was emitted")
```

Also, not sure if this is an issue, but you don't have the slot decorator on your update function


```python
@Slot(np.ndarray)
def update(self, data: np.ndrarray):
    self.img.setImage(data, levels=(0, 1))
    # self.graphicsView.show()  # I don't this this is necessary provided it was already showing
```

**Comment by kevinNejad** [comment]
The project is very small, it's kind of a test for now, but good tips.
When I add decorator(`@pyqtSlot(np.ndarray)`) I get this error  `TypeError: connect() failed between Model.sigNewData[numpy.ndarray] and update()`
Seems like I can't emit a signal inside a for loop. I don't know why though. Basically train methods emit signal with new data and plot must get new data and update the image.

**Comment by 2xB** [comment]
That's already great! You're nearly done. Your issue is that everything is currently running in a single process and a single thread, meaning while you execute your for loop, the GUI is frozen, because it can't update. An easy and pretty solution is to make `Model` a QThread. Here is an example of how that would look like:

View.py unchanged.

Model.py:
```python3
import numpy as np
from PyQt5.QtCore import pyqtSignal, QThread

class Model(QThread):
    def __init__(self):
        QThread.__init__(self)  

    signal= pyqtSignal(np.ndarray)
    def run(self):
        for step in range(1000):
            dummy_data = np.random.normal(size=(100,100))
            self.signal.emit(dummy_data)
```

controller.py:
```python3
from model import Model
from view import View
import numpy as np
import sys

from PyQt5 import QtCore, QtGui, QtWidgets

class Controller:
    def __init__(self):
        self.model = Model()
        self.view = View(self)

if __name__=='__main__':

    ctrl = Controller()

    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    
    ui = ctrl.view
    ui.setupUi(MainWindow)
    MainWindow.show()

    # w = ClassificationPlot(ui.TaskPlot)
    ctrl.model.signal.connect(ui.update)
    
    ctrl.model.start()

    sys.exit(app.exec_())
```

**Comment by kevinNejad** [comment]
You are such a life saver.
The library is awesome and kind of addictive. 
Thank you so much.
I close the issue then.
